File: ktox-lua-cc/src/main/kotlin/programs/ExcavatePro.kt
package programs

import common.turtleGetItemName
import common.turtleDigUp
import common.turtleGetFuelLevel
import common.turtleGetItemCount
import common.turtlePlaceUp
import common.turtleSelect
import common.turtleTransferTo
import lib.CHEST_INTAKE_SLOT
import lib.IntSpan
import lib.Movement
import lib.calibrateMovement
import lib.cargoFull
import lib.dumpCargo
import lib.gpsLocate
import lib.headingDx
import lib.headingDz
import lib.navigateTo
import lib.pastEnd
import lib.restockChest
import lib.stepFor

// Straight-down excavator (like CC's built-in excavate) that also lines
// the shaft with a spiral staircase and torches as it descends.
//
// Usage: excavatepro <width> [<length>] [<depth>]
//   - width: footprint size along the turtle's right-hand side at start.
//   - length (optional, defaults to width — square footprint): footprint
//     size along the turtle's forward direction at start.
//   - depth (optional, blocks below home's Y — always positive): stop
//     once this many levels below the start have been dug. Omit to dig
//     until bedrock (a fresh digDown() attempt fails). Relative to home
//     rather than an absolute world Y, so this works identically with or
//     without GPS (see calibrateMovement() in lib/Movement.kt).
//   Positional convention: arg COUNT decides meaning (1 = width only,
//   2 = width+length, 3 = width+length+depth) since there's no way to
//   tell a bare length from a bare depth apart otherwise. To request a
//   square hole to a specific depth, pass width twice: "5 5 50".
//
// Geometry: one step (cobblestone — see below) or, every TORCH_INTERVAL
// levels, one torch instead, is placed each Y-level at a position that
// walks around the footprint's perimeter, so the full descent traces one
// connected spiral (wrapping around for as many laps as the depth
// needs). It's placed into the CEILING of the layer just finished — i.e.
// the floor of the layer above, which was already fully cleared on the
// previous iteration — via turtlePlaceUp(), not down into ground the
// turtle is about to descend through. That means no bookkeeping is
// needed to protect it from the next layer's sweep: the layer it's
// placed into has already been swept and is never revisited, so there's
// nothing left to accidentally re-dig it. This whole placement scheme is
// the least-tested part of this program (no way to exercise real turtle
// placement outside a live world).
//
// Steps are plain cobblestone, not actual stair blocks: mining through
// stone always produces it, so it's entirely self-sustaining (no chest
// trip needed), and unlike a real stair block it has no facing to get
// placed wrong. It won't look like a proper sloped staircase, just
// ascending platforms, but it's a reliable climbable path.
//
// Supply: a chest at the turtle's start position holds charcoal and
// torches ("minecraft:torch") mixed together — sucked one item at a
// time into CHEST_INTAKE_SLOT and sorted by name. TORCH_SLOT holds torch
// stock; cobblestone is never dumped as cargo (kept for steps, see
// shouldKeepSlot); everything else is general cargo, dumped to overflow
// chests. Chest logic itself (dumping, restocking, the fuel allowlist)
// is shared with Digsite via lib/Chest.kt — this file only decides
// which slots to keep off-limits and what to do with a non-fuel item.

const val EP_FUEL_SAFETY_MARGIN = 20
const val TORCH_INTERVAL = 6
const val TORCH_SLOT = 3
const val RESTOCK_ATTEMPTS = 32
const val COBBLESTONE_RESERVE = 64

fun main(args: Array<String>) {
    if (args.size < 1) {
        println("Usage: excavatepro <width> (<length>) (<depth>)")
        return
    }

    val width = args[1].toInt()
    val length = if (args.size >= 2) args[2].toInt() else width
    val hasDepth = args.size >= 3
    val depth = if (hasDepth) args[3].toInt() else 0

    println("Calibrating position via GPS...")
    val movement = calibrateMovement()
    if (movement.gpsEnabled) {
        println("Home at x=${movement.homeX} y=${movement.homeY} z=${movement.homeZ}")
    } else {
        println("No GPS available - running on dead reckoning only (no drift checks).")
    }

    // Relative to homeY rather than an absolute world Y, so this stop
    // condition means the same thing regardless of GPS availability.
    val yTarget = movement.homeY - depth

    val fwdDx = headingDx(movement.homeHeading)
    val fwdDz = headingDz(movement.homeHeading)
    val rightHeading = (movement.homeHeading + 1) % 4
    val rgtDx = headingDx(rightHeading)
    val rgtDz = headingDz(rightHeading)

    val xOther = movement.homeX + rgtDx * (width - 1) + fwdDx * (length - 1)
    val zOther = movement.homeZ + rgtDz * (width - 1) + fwdDz * (length - 1)
    val xSpan = IntSpan(movement.homeX, xOther)
    val zSpan = IntSpan(movement.homeZ, zOther)

    // Used only as a return-path waypoint (see navigateWithRecovery) —
    // every layer's digLayer sweep clears it along with everything else,
    // so it's guaranteed open at every height above wherever the turtle
    // currently is, unlike a direct path to home which could run into
    // this layer's own placed step/torch or, at the very bottom, bedrock
    // on multiple sides at once.
    val centerX = movement.homeX + rgtDx * (width / 2) + fwdDx * (length / 2)
    val centerZ = movement.homeZ + rgtDz * (width / 2) + fwdDz * (length / 2)

    val hasStaircase = width >= 2 && length >= 2

    var y = movement.homeY
    var step = 0
    var levelsSinceTorch = 0
    var digging = true

    while (digging) {
        epEnsureFuelAndSpace(movement, centerX, centerZ)
        var blocked = false
        if (!digLayer(movement, xSpan, zSpan, y, centerX, centerZ)) {
            blocked = true
        }

        // Nothing is placed at or above homeY: the topmost layer has
        // nothing pre-cleared above it to place into (there was no
        // previous iteration), and more importantly homeY is the fuel
        // chest's level — every overflow chest sits along exactly the
        // same right-hand line (dz=0) that segment 1 of the perimeter
        // walk below would otherwise place its first few steps/torches
        // into, burying the chest row (or home itself, at cell (0,0))
        // under a placed block. The first real step goes in once
        // excavation reaches the layer strictly below homeY.
        if (!blocked && hasStaircase && y + 1 < movement.homeY) {
            val cell = perimeterCell(step, width, length)
            // cell.dx/dz are offsets along the right/forward axes (the same
            // basis used to build xSpan/zSpan above) — NOT along x/z
            // directly. Which world axis "right" and "forward" map to
            // depends on the turtle's starting heading (north/south: right
            // is x, forward is z; east/west: the other way around), so
            // reusing xSpan/zSpan's step values here would silently swap
            // width and length whenever the turtle started facing east or
            // west. rgtDx/rgtDz/fwdDx/fwdDz already encode that mapping
            // correctly — reuse them instead.
            val cellX = movement.homeX + rgtDx * cell.dx + fwdDx * cell.dz
            val cellZ = movement.homeZ + rgtDz * cell.dx + fwdDz * cell.dz
            if (!navigateTo(movement, cellX, y, cellZ)) {
                blocked = true
            } else {
                // The step is placed on every transition, unconditionally
                // - it's the only thing making that transition climbable.
                // A torch used to be placed here INSTEAD of the step every
                // TORCH_INTERVAL layers, which left a gap in the staircase
                // at every one of those transitions (a torch isn't
                // something you can stand on). Torches are now a bonus,
                // placed at a separate cell below instead of competing
                // for this one.
                val stepSlot = findCobblestoneSlot()
                if (stepSlot == -1) {
                    println("No cobblestone on hand for a step at y=${y} - skipping this one.")
                } else {
                    turtleDigUp()
                    turtleSelect(stepSlot)
                    turtlePlaceUp()
                }

                levelsSinceTorch += 1
                if (levelsSinceTorch >= TORCH_INTERVAL) {
                    // Next perimeter cell over — not due for its own
                    // (real) step until the following, one-lower
                    // iteration, so there's no Y-level collision with it,
                    // and it's right next to the step just placed.
                    val torchCell = perimeterCell(step + 1, width, length)
                    val torchX = movement.homeX + rgtDx * torchCell.dx + fwdDx * torchCell.dz
                    val torchZ = movement.homeZ + rgtDz * torchCell.dx + fwdDz * torchCell.dz
                    if (!navigateTo(movement, torchX, y, torchZ)) {
                        println("Couldn't reach a torch position at y=${y} - skipping this one.")
                    } else {
                        ensureTorchSupply(movement, centerX, centerZ)
                        turtleDigUp()
                        turtleSelect(TORCH_SLOT)
                        turtlePlaceUp()
                    }
                    levelsSinceTorch = 0
                }
                step += 1
            }
        }

        if (blocked) {
            // Movement genuinely can't proceed — most likely bedrock (dig
            // succeeds into a pocket between two bedrock blocks, but the
            // move into the next one still fails). Stop and fall through
            // to the return-home below rather than retrying a move that
            // will just fail again.
            println("Movement blocked at y=${y} (bedrock or another undiggable obstruction) - stopping.")
            digging = false
        } else if (hasDepth && y <= yTarget) {
            digging = false
        } else {
            val moved = movement.down()
            y = movement.y
            if (!moved) {
                println("Hit bedrock (or an undiggable block) at y=${y}.")
                digging = false
            }
        }
    }

    println("Excavation complete. Returning home...")
    if (!navigateWithRecovery(movement, centerX, centerZ)) {
        println("Could not reach the center column even after climbing - staying put.")
    } else {
        // From the center column, everything above was already fully
        // cleared by earlier digLayer calls, so a plain climb-then-walk
        // is safe the rest of the way.
        navigateTo(movement, centerX, movement.homeY, centerZ)
        navigateTo(movement, movement.homeX, movement.homeY, movement.homeZ)
        movement.faceHeading(movement.homeHeading)
        println("Home.")
    }
}

const val MAX_RECOVERY_ASCENTS = 32

// Tries to reach (targetX, targetZ) at the turtle's current height. If
// blocked — most likely wedged in a bedrock pocket, possibly boxed in on
// more than one side — climbs one block (retracing already-dug
// territory, since everything above the current layer was already fully
// swept) and retries at the new height, repeating until it succeeds or
// MAX_RECOVERY_ASCENTS is hit. Gives up (returns false) immediately if
// even climbing fails — that would mean something is blocking upward
// too, which shouldn't happen given every layer above is already clear,
// but there's no sense retrying a move that can't work.
fun navigateWithRecovery(m: Movement, targetX: Int, targetZ: Int): Boolean {
    if (navigateTo(m, targetX, m.y, targetZ)) {
        return true
    }
    var attempts = 0
    var reached = false
    while (!reached && attempts < MAX_RECOVERY_ASCENTS) {
        if (!m.up()) {
            return false
        }
        reached = navigateTo(m, targetX, m.y, targetZ)
        attempts += 1
    }
    return reached
}

// -- Perimeter walk --
//
// Cell(0,0) is the turtle's starting corner. Walks clockwise: along the
// "width" edge, then the "length" edge, back along "width", back along
// "length" — wrapping via modulo for however many laps the depth needs.
// Only valid for width >= 2 and length >= 2 (checked by the caller).

data class Cell(val dx: Int, val dz: Int)

fun perimeterLength(width: Int, length: Int): Int {
    return 2 * width + 2 * length - 4
}

fun perimeterCell(step: Int, width: Int, length: Int): Cell {
    val p = perimeterLength(width, length)
    var i = step % p
    if (i < width) {
        return Cell(i, 0)
    }
    i -= width
    if (i < length - 1) {
        return Cell(width - 1, i + 1)
    }
    i -= (length - 1)
    if (i < width - 1) {
        return Cell(width - 2 - i, length - 1)
    }
    i -= (width - 1)
    return Cell(0, length - 2 - i)
}

// -- Full-layer excavation --

// Returns false the moment any move is genuinely blocked (see
// navigateTo) — the layer may be left partially swept when that happens;
// the caller decides how to react (main() stops and returns home).
fun digLayer(m: Movement, xSpan: IntSpan, zSpan: IntSpan, y: Int, centerX: Int, centerZ: Int): Boolean {
    val xStep = stepFor(xSpan)
    val zStep = stepFor(zSpan)
    var x = xSpan.start
    var forwardZ = true
    var ok = true
    while (ok && !pastEnd(x, xSpan.finish, xStep)) {
        if (forwardZ) {
            var z = zSpan.start
            while (ok && !pastEnd(z, zSpan.finish, zStep)) {
                epEnsureFuelAndSpace(m, centerX, centerZ)
                if (!navigateTo(m, x, y, z)) {
                    ok = false
                }
                z += zStep
            }
        } else {
            var z = zSpan.finish
            while (ok && !pastEnd(z, zSpan.start, -zStep)) {
                epEnsureFuelAndSpace(m, centerX, centerZ)
                if (!navigateTo(m, x, y, z)) {
                    ok = false
                }
                z -= zStep
            }
        }
        forwardZ = !forwardZ
        x += xStep
    }
    return ok
}

// -- Base servicing (fuel + item disposal + torch restock) --

fun isTorchItem(name: String): Boolean {
    return name == "minecraft:torch"
}

// True if this slot's cobblestone falls within the first
// COBBLESTONE_RESERVE items counted across cobblestone-holding slots in
// order from slot 1 - roughly "the first stack". Mining constantly
// produces more cobblestone than a staircase actually needs, and it used
// to be kept in EVERY slot that held any, uncapped - cargo filled with
// cobblestone alone, triggering base trips (see epServiceAtBase) far
// more often than necessary. Recomputed fresh each call (a plain O(slot)
// scan) rather than tracked as running state, since shouldKeepSlot is
// called independently per slot by the shared cargoFull/dumpCargo loops
// in lib/Chest.kt with no state passed between calls.
fun withinCobblestoneReserve(slot: Int): Boolean {
    var s = 1
    var totalBefore = 0
    while (s < slot) {
        val name = turtleGetItemName(s)
        if (name != null && name == "minecraft:cobblestone") {
            totalBefore += turtleGetItemCount(s)
        }
        s += 1
    }
    return totalBefore < COBBLESTONE_RESERVE
}

// A slot's current content is worth keeping — never dumped as cargo —
// when: it's the transient intake slot and happens to be empty right
// now; it's the torch slot and actually holds a torch; or it holds
// cobblestone within the reserve (see withinCobblestoneReserve) — cargo
// beyond that reserve is treated as ordinary cargo like anything else.
// Checking by content rather than just slot number matters: a slot that
// doesn't hold its intended item (e.g. TORCH_SLOT holding ordinary mined
// cobblestone, because it happened to be the first empty slot the very
// first time the turtle mined something, before any chest visit ever
// occurred) is NOT protected — it's ordinary cargo that needs to get
// dumped so the slot is actually empty for the real item next visit.
fun shouldKeepSlot(slot: Int): Boolean {
    if (slot == CHEST_INTAKE_SLOT) {
        return turtleGetItemCount(slot) == 0
    }
    val name = turtleGetItemName(slot)
    if (name == null) {
        return slot == TORCH_SLOT
    }
    if (slot == TORCH_SLOT) {
        return isTorchItem(name)
    }
    if (name == "minecraft:cobblestone") {
        return withinCobblestoneReserve(slot)
    }
    return false
}

fun epNeedsService(m: Movement): Boolean {
    val fuel = turtleGetFuelLevel()
    val distance = m.distanceHome()
    if (fuel < distance + EP_FUEL_SAFETY_MARGIN) {
        return true
    }
    return cargoFull { slot -> shouldKeepSlot(slot) }
}

fun epEnsureFuelAndSpace(m: Movement, centerX: Int, centerZ: Int) {
    if (epNeedsService(m)) {
        epServiceAtBase(m, centerX, centerZ)
    }
}

fun ensureTorchSupply(m: Movement, centerX: Int, centerZ: Int) {
    val name = turtleGetItemName(TORCH_SLOT)
    val hasTorch = name != null && isTorchItem(name)
    if (!hasTorch) {
        epServiceAtBase(m, centerX, centerZ)
    }
}

// Searches inventory for a slot already holding cobblestone (from mining
// through stone) to use as step material. Returns -1 if the turtle
// genuinely has none right now — rare, but possible early in a dig or in
// a stone-poor area.
fun findCobblestoneSlot(): Int {
    var slot = 1
    var found = -1
    while (slot <= 16 && found == -1) {
        if (turtleGetItemCount(slot) > 0) {
            val name = turtleGetItemName(slot)
            if (name != null && name == "minecraft:cobblestone") {
                found = slot
            }
        }
        slot += 1
    }
    return found
}

fun epServiceAtBase(m: Movement, centerX: Int, centerZ: Int) {
    println("Returning to base to refuel/dump inventory/restock supplies...")
    val returnX = m.x
    val returnY = m.y
    val returnZ = m.z

    // Route vertical travel through the center column - guaranteed clear
    // of placed steps at every height (see centerX/centerZ in main()) -
    // rather than climbing/descending straight from wherever the turtle
    // happens to be, which could dig straight through (destroy) a step
    // placed on an earlier, higher layer.
    navigateTo(m, centerX, m.y, centerZ)
    navigateTo(m, centerX, m.homeY, centerZ)

    dumpCargo(m) { slot -> shouldKeepSlot(slot) }
    restockAtChest(m)

    navigateTo(m, centerX, m.homeY, centerZ)
    navigateTo(m, centerX, returnY, centerZ)
    navigateTo(m, returnX, returnY, returnZ)

    // Drift safety check after the round trip — only meaningful when GPS
    // was available at calibration; otherwise there's no ground truth to
    // check against.
    if (m.gpsEnabled) {
        val checked = gpsLocate()
        if (checked != null) {
            m.x = checked.x
            m.y = checked.y
            m.z = checked.z
        }
    }
    println("Resuming excavation.")
}

// Anything restockChest's fuel allowlist doesn't already claim — a
// torch goes to TORCH_SLOT; anything else is left for restockChest to
// drop back and stop on. Written with an explicit `handled` local, not
// as an if/else expression — see AGENTS.md on why a multi-statement
// if/else branch used as a lambda's implicit return can transpile to
// invalid Lua.
fun restockAtChest(m: Movement) {
    restockChest(m, RESTOCK_ATTEMPTS) { name ->
        var handled = false
        if (isTorchItem(name)) {
            turtleSelect(CHEST_INTAKE_SLOT)
            // Checked, not assumed: turtleTransferTo() returns false (and
            // moves nothing) if TORCH_SLOT is occupied by something that
            // can't stack with a torch. Reporting handled=true anyway
            // would silently strand the torch in CHEST_INTAKE_SLOT, where
            // it'd eventually get swept up and dumped as ordinary cargo
            // on the next base trip instead of ever reaching TORCH_SLOT.
            handled = turtleTransferTo(TORCH_SLOT, 64)
        }
        handled
    }
}

File: src/pkg/player/8durt/ExcavatePro.lua
-- package: programs

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "ExcavatePro.kt", {["1-21"]=1,["22"]=77,["23"]=78,["24-25"]=79,["26"]=82,["27"]=83,["28"]=84,["29"]=85,["30"]=87,["31"]=88,["32"]=89,["33-34"]=90,["35-36"]=92,["37"]=97,["38"]=99,["39"]=100,["40"]=101,["41"]=102,["42"]=103,["43"]=105,["44"]=106,["45"]=107,["46"]=108,["47"]=129,["48"]=130,["49"]=131,["50"]=132,["51"]=134,["52"]=136,["53"]=137,["54"]=138,["55"]=139,["56"]=141,["57"]=142,["58"]=143,["59"]=144,["60-61"]=145,["62"]=157,["63"]=158,["64"]=168,["65"]=169,["66"]=170,["67-68"]=171,["69"]=181,["70"]=182,["71-72"]=183,["73"]=185,["74"]=186,["75-76"]=187,["77"]=190,["78"]=191,["79"]=196,["80"]=197,["81"]=198,["82"]=199,["83-84"]=200,["85"]=202,["86"]=203,["87"]=204,["88-89"]=205,["90-91"]=207,["92-94"]=209,["95"]=213,["96"]=219,["97-98"]=220,["99-100"]=222,["101"]=224,["102"]=225,["103"]=226,["104"]=227,["105-108"]=228,["109"]=233,["110"]=234,["111-112"]=235,["113"]=240,["114"]=241,["115"]=242,["116-126"]=243,["127"]=259,["128-129"]=260,["130"]=262,["131"]=263,["132"]=264,["133"]=265,["134-135"]=266,["136"]=268,["137-138"]=269,["139-178"]=271,["179-186"]=284,["187"]=288,["188"]=289,["189"]=290,["190-191"]=291,["192"]=293,["193"]=294,["194-195"]=295,["196"]=297,["197"]=298,["198-199"]=299,["200"]=301,["201-211"]=302,["212"]=311,["213"]=312,["214"]=313,["215"]=314,["216"]=315,["217"]=316,["218"]=317,["219"]=318,["220"]=319,["221"]=320,["222"]=321,["223-224"]=322,["225-227"]=324,["228"]=327,["229"]=328,["230"]=329,["231"]=330,["232-233"]=331,["234-236"]=333,["237"]=336,["238-239"]=337,["240-245"]=339,["246-251"]=345,["252"]=359,["253"]=360,["254"]=361,["255"]=362,["256"]=363,["257-258"]=364,["259-260"]=366,["261-266"]=368,["267"]=383,["268-269"]=384,["270"]=386,["271"]=387,["272-273"]=388,["274"]=390,["275-276"]=391,["277"]=393,["278-279"]=394,["280-285"]=396,["286"]=400,["287"]=401,["288"]=402,["289-290"]=403,["291-299"]=405,["300"]=409,["301-308"]=410,["309"]=415,["310"]=416,["311"]=417,["312-317"]=418,["318"]=427,["319"]=428,["320"]=429,["321"]=430,["322"]=431,["323"]=432,["324-326"]=433,["327-328"]=436,["329-335"]=438,["336"]=442,["337"]=443,["338"]=444,["339"]=445,["340"]=452,["341"]=453,["342-344"]=455,["345"]=456,["346"]=458,["347"]=459,["348"]=460,["349"]=465,["350"]=466,["351"]=467,["352"]=468,["353"]=469,["354-356"]=470,["357-361"]=473,["362-371"]=483}, "programs")
ktox_require("lib/Chest")
ktox_require("lib/Span")
ktox_require("lib/Movement")
ktox_require("lib/Position")

EP_FUEL_SAFETY_MARGIN = 20

TORCH_INTERVAL = 6

TORCH_SLOT = 3

RESTOCK_ATTEMPTS = 32

COBBLESTONE_RESERVE = 64

---@param args table
function main(args)
    if #(args) < 1 then
        println("Usage: excavatepro <width> (<length>) (<depth>)")
        return
    end
    local width = ktox_toInt(args[1])
    local length = (#(args) >= 2 and ktox_toInt(args[2]) or width)
    local hasDepth = #(args) >= 3
    local depth = (hasDepth and ktox_toInt(args[3]) or 0)
    println("Calibrating position via GPS...")
    local movement = calibrateMovement()
    if movement.gpsEnabled then
        println("Home at x=" .. tostring(movement.homeX) .. " y=" .. tostring(movement.homeY) .. " z=" .. tostring(movement.homeZ))
    else
        println("No GPS available - running on dead reckoning only (no drift checks).")
    end
    local yTarget = movement.homeY - depth
    local fwdDx = headingDx(movement.homeHeading)
    local fwdDz = headingDz(movement.homeHeading)
    local rightHeading = (movement.homeHeading + 1) % 4
    local rgtDx = headingDx(rightHeading)
    local rgtDz = headingDz(rightHeading)
    local xOther = movement.homeX + rgtDx * (width - 1) + fwdDx * (length - 1)
    local zOther = movement.homeZ + rgtDz * (width - 1) + fwdDz * (length - 1)
    local xSpan = IntSpan:new(movement.homeX, xOther)
    local zSpan = IntSpan:new(movement.homeZ, zOther)
    local halfWidth = (width - width % 2) / 2
    local halfLength = (length - length % 2) / 2
    local centerX = movement.homeX + rgtDx * halfWidth + fwdDx * halfLength
    local centerZ = movement.homeZ + rgtDz * halfWidth + fwdDz * halfLength
    local hasStaircase = width >= 2 and length >= 2
    local y = movement.homeY
    local step = 0
    local levelsSinceTorch = 0
    local digging = true
    while digging do
        epEnsureFuelAndSpace(movement, centerX, centerZ)
        local blocked = false
        if not digLayer(movement, xSpan, zSpan, y, centerX, centerZ) then
            blocked = true
        end
        if not blocked and hasStaircase and y + 1 < movement.homeY then
            local cell = perimeterCell(step, width, length)
            local cellX = movement.homeX + rgtDx * cell.dx + fwdDx * cell.dz
            local cellZ = movement.homeZ + rgtDz * cell.dx + fwdDz * cell.dz
            if not navigateTo(movement, cellX, y, cellZ) then
                blocked = true
            else
                local stepSlot = findCobblestoneSlot()
                if stepSlot == -1 then
                    println("No cobblestone on hand for a step at y=" .. tostring(y) .. " - skipping this one.")
                else
                    turtle.digUp()
                    turtle.select(stepSlot)
                    turtle.placeUp()
                end
                levelsSinceTorch = ktox_plusAssign(levelsSinceTorch, 1)
                if levelsSinceTorch >= TORCH_INTERVAL then
                    local torchCell = perimeterCell(step + 1, width, length)
                    local torchX = movement.homeX + rgtDx * torchCell.dx + fwdDx * torchCell.dz
                    local torchZ = movement.homeZ + rgtDz * torchCell.dx + fwdDz * torchCell.dz
                    if not navigateTo(movement, torchX, y, torchZ) then
                        println("Couldn\'t reach a torch position at y=" .. tostring(y) .. " - skipping this one.")
                    else
                        ensureTorchSupply(movement, centerX, centerZ)
                        turtle.digUp()
                        turtle.select(TORCH_SLOT)
                        turtle.placeUp()
                    end
                    levelsSinceTorch = 0
                end
                step = ktox_plusAssign(step, 1)
            end
        end
        if blocked then
            println("Movement blocked at y=" .. tostring(y) .. " (bedrock or another undiggable obstruction) - stopping.")
            digging = false
        elseif hasDepth and y <= yTarget then
            digging = false
        else
            local moved = movement:down()
            y = movement.y
            if not moved then
                println("Hit bedrock (or an undiggable block) at y=" .. tostring(y) .. ".")
                digging = false
            end
        end
    end
    println("Excavation complete. Returning home...")
    if not navigateWithRecovery(movement, centerX, centerZ) then
        println("Could not reach the center column even after climbing - staying put.")
    else
        navigateTo(movement, centerX, movement.homeY, centerZ)
        navigateTo(movement, movement.homeX, movement.homeY, movement.homeZ)
        movement:faceHeading(movement.homeHeading)
        println("Home.")
    end
end

MAX_RECOVERY_ASCENTS = 32

---@param m Movement
---@param targetX number
---@param targetZ number
---@return boolean
function navigateWithRecovery(m, targetX, targetZ)
    if navigateTo(m, targetX, m.y, targetZ) then
        return true
    end
    local attempts = 0
    local reached = false
    while not reached and attempts < MAX_RECOVERY_ASCENTS do
        if not m:up() then
            return false
        end
        reached = navigateTo(m, targetX, m.y, targetZ)
        attempts = ktox_plusAssign(attempts, 1)
    end
    return reached
end

---@class Cell
---@field dx number
---@field dz number
Cell = {}
Cell.__index = Cell

function Cell:new(dx, dz)
    local self = setmetatable({}, Cell)
    self.dx = dx
    self.dz = dz
    return self
end

function Cell:equals(other)
    return self.dx == other.dx and self.dz == other.dz
end
Cell.__eq = function(a, b) return a:equals(b) end
function Cell:toString()
    return "Cell(" .. "dx=" .. tostring(self.dx) .. ", " .. "dz=" .. tostring(self.dz) .. ")"
end
Cell.__tostring = function(a) return a:toString() end
function Cell:copy(dx, dz)
    if dx == nil then dx = self.dx end
    if dz == nil then dz = self.dz end
    return Cell:new(dx, dz)
end
function Cell:component1()
    return self.dx
end
function Cell:component2()
    return self.dz
end

---@param width number
---@param length number
---@return number
function perimeterLength(width, length)
    return 2 * width + 2 * length - 4
end

---@param step number
---@param width number
---@param length number
---@return Cell
function perimeterCell(step, width, length)
    local p = perimeterLength(width, length)
    local i = step % p
    if i < width then
        return Cell:new(i, 0)
    end
    i = ktox_minusAssign(i, width)
    if i < length - 1 then
        return Cell:new(width - 1, i + 1)
    end
    i = ktox_minusAssign(i, (length - 1))
    if i < width - 1 then
        return Cell:new(width - 2 - i, length - 1)
    end
    i = ktox_minusAssign(i, (width - 1))
    return Cell:new(0, length - 2 - i)
end

---@param m Movement
---@param xSpan IntSpan
---@param zSpan IntSpan
---@param y number
---@param centerX number
---@param centerZ number
---@return boolean
function digLayer(m, xSpan, zSpan, y, centerX, centerZ)
    local xStep = stepFor(xSpan)
    local zStep = stepFor(zSpan)
    local x = xSpan.start
    local forwardZ = true
    local ok = true
    while ok and not pastEnd(x, xSpan.finish, xStep) do
        if forwardZ then
            local z = zSpan.start
            while ok and not pastEnd(z, zSpan.finish, zStep) do
                epEnsureFuelAndSpace(m, centerX, centerZ)
                if not navigateTo(m, x, y, z) then
                    ok = false
                end
                z = ktox_plusAssign(z, zStep)
            end
        else
            local z = zSpan.finish
            while ok and not pastEnd(z, zSpan.start, -zStep) do
                epEnsureFuelAndSpace(m, centerX, centerZ)
                if not navigateTo(m, x, y, z) then
                    ok = false
                end
                z = ktox_minusAssign(z, zStep)
            end
        end
        forwardZ = not forwardZ
        x = ktox_plusAssign(x, xStep)
    end
    return ok
end

---@param name string
---@return boolean
function isTorchItem(name)
    return name == "minecraft:torch"
end

---@param slot number
---@return boolean
function withinCobblestoneReserve(slot)
    local s = 1
    local totalBefore = 0
    while s < slot do
        local name = ktoxGetItemName(s)
        if name ~= nil and name == "minecraft:cobblestone" then
            totalBefore = ktox_plusAssign(totalBefore, turtle.getItemCount(s))
        end
        s = ktox_plusAssign(s, 1)
    end
    return totalBefore < COBBLESTONE_RESERVE
end

---@param slot number
---@return boolean
function shouldKeepSlot(slot)
    if slot == CHEST_INTAKE_SLOT then
        return turtle.getItemCount(slot) == 0
    end
    local name = ktoxGetItemName(slot)
    if name == nil then
        return slot == TORCH_SLOT
    end
    if slot == TORCH_SLOT then
        return isTorchItem(name)
    end
    if name == "minecraft:cobblestone" then
        return withinCobblestoneReserve(slot)
    end
    return false
end

---@param m Movement
---@return boolean
function epNeedsService(m)
    local fuel = turtle.getFuelLevel()
    local distance = m:distanceHome()
    if fuel < distance + EP_FUEL_SAFETY_MARGIN then
        return true
    end
    return cargoFull(function(slot)
        return shouldKeepSlot(slot)
    end)
end

---@param m Movement
---@param centerX number
---@param centerZ number
function epEnsureFuelAndSpace(m, centerX, centerZ)
    if epNeedsService(m) then
        epServiceAtBase(m, centerX, centerZ)
    end
end

---@param m Movement
---@param centerX number
---@param centerZ number
function ensureTorchSupply(m, centerX, centerZ)
    local name = ktoxGetItemName(TORCH_SLOT)
    local hasTorch = name ~= nil and isTorchItem(name)
    if not hasTorch then
        epServiceAtBase(m, centerX, centerZ)
    end
end

---@return number
function findCobblestoneSlot()
    local slot = 1
    local found = -1
    while slot <= 16 and found == -1 do
        if turtle.getItemCount(slot) > 0 then
            local name = ktoxGetItemName(slot)
            if name ~= nil and name == "minecraft:cobblestone" then
                found = slot
            end
        end
        slot = ktox_plusAssign(slot, 1)
    end
    return found
end

---@param m Movement
---@param centerX number
---@param centerZ number
function epServiceAtBase(m, centerX, centerZ)
    println("Returning to base to refuel/dump inventory/restock supplies...")
    local returnX = m.x
    local returnY = m.y
    local returnZ = m.z
    navigateTo(m, centerX, m.y, centerZ)
    navigateTo(m, centerX, m.homeY, centerZ)
    dumpCargo(m, function(slot)
        return shouldKeepSlot(slot)
    end)
    restockAtChest(m)
    navigateTo(m, centerX, m.homeY, centerZ)
    navigateTo(m, centerX, returnY, centerZ)
    navigateTo(m, returnX, returnY, returnZ)
    if m.gpsEnabled then
        local checked = gpsLocate()
        if checked ~= nil then
            m.x = checked.x
            m.y = checked.y
            m.z = checked.z
        end
    end
    println("Resuming excavation.")
end

---@param m Movement
function restockAtChest(m)
    restockChest(m, RESTOCK_ATTEMPTS, function(name)
        local handled = false
        if isTorchItem(name) then
            turtle.select(CHEST_INTAKE_SLOT)
            handled = turtle.transferTo(TORCH_SLOT, 64)
        end
        return handled
    end)
end


main({...})
